google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractResultSet.java

```diff
@@ -778,6 +827,52 @@ protected List<String> getJsonListInternal(int columnIndex) {
       return Collections.unmodifiableList((List<String>) rowData.get(columnIndex));
     }
 
+    @Override
+    @SuppressWarnings("unchecked") // We know ARRAY<PROTO> produces a List<ByteArray>.
+    protected <T extends AbstractMessage> List<T> getProtoMessageListInternal(
+        int columnIndex, T message) {
+      Preconditions.checkNotNull(
+          message,
+          "Proto message may not be null. Use MyProtoClass.getDefaultInstance() as a parameter value.");
+
+      List<ByteArray> bytesArray = (List<ByteArray>) rowData.get(columnIndex);
+
+      try {
+        List<T> protoMessagesList = new ArrayList<>(bytesArray.size());
+        for (ByteArray protoMessageBytes : bytesArray) {
+          if (protoMessageBytes == null) {
+            protoMessagesList.add(null);
+          } else {
+            protoMessagesList.add(
+                (T) message.toBuilder().mergeFrom(protoMessageBytes.toByteArray()).build());
+          }
+        }
+        return protoMessagesList;
+      } catch (InvalidProtocolBufferException e) {
+        throw SpannerExceptionFactory.asSpannerException(e);
+      }
+    }
+
+    @Override
+    @SuppressWarnings("unchecked") // We know ARRAY<ENUM> produces a List<Long>.
+    protected <T extends ProtocolMessageEnum> List<T> getProtoEnumListInternal(
+        int columnIndex, Function<Integer, ProtocolMessageEnum> method) {
+      Preconditions.checkNotNull(
+          method, "Method may not be null. Use 'MyProtoEnum::forNumber' as a parameter value.");
```

Comment:
@Shua1: I prefer getProtoEnumListInternal(idx, my_method) here. Adding `Function<Integer, ProtocolMessageEnum>` method as a parameter makes more sense to me, because implementation for `forNumber` will be specific to the `ProtocolMessageEnum` class.

With getProtoEnumListInternal(idx), the Java client will not have any way to convert `Long` from Spanner backend into corresponding enum.